### ISSUE TITLE
Calendar button stays disabled

### ISSUE THREAD

**Issue by hanslammerts** [Offerel/roundcube_caldavzap#4]
Hi,

First of all, thanks for the plugin. I'm hoping to get it working, but wasn't succesful so far.
Here's what I did:

1) Installed Roundcube (1.4), which is working as far as email concerns.
2) Installed Davical 1.1.7, and added a principal which is the same user/password I use to login to Roundcube. Calendars are reachable after logging in at URI /davical/caldav.php/
3) Installed the caldavzap Roundcube plugin, and configured both $config['caldav_server'] = 'localhost' , and $config['caldav_path'] = '/davical/caldav.php'
4) Enabled the plugin in the Roundcube config.inc.php

Now, when I login to Roundcube, I do see a button Calendar, but it is greyed out, and not selectable (disabled).
There are no errors in any logs whatsoever.

Can you please explain why the calendar button remains disabled, please ?

Regards,
Hans

**Comment by Offerel** [comment]
If I understand you right, you mean that the button in the Roundcube taskbar is greyed out and disabled? Or do you mean a button in the calendar plugin page itself?

Additionally you should see all requests at least in your webserver access log and if there are problems, they should appear in the error log (eg. failed logins for example).

Additionally, when you configure the domain with 'localhost', this means accessing the calendar server with port 80 and no SSL. If this is your setup, this should work, I would only give you a small hint about this.

**Comment by hanslammerts** [comment]
Thanks for responding this fast !
It is the button in Roundcube that is disabled, not a button on the calendar page itself. I can't even get to the calendar page because the button "Calendar" is disabled.......
And yes, Davical is reachable on post 80, localhost...

I'm guessing I do not see anything in the access/error log because the button does not send me anywhere.....

**Comment by Offerel** [comment]
Yes, I fear you are right. The plug-in try only to login at the calendar, if the 'task' is started. I have no idea, why the button could be deactivated. There is no such disable function build in to this plugin, at least not for the button in the taskbar.

Do you have other plugins activated in your installation and if so, can please disable them temporary? Maybe there is a css style, or JavaScript function which disable the button.

Do you use larry or the classic skin in Roundcube?

**Comment by hanslammerts** [comment]
Update:
Somehow I installed an older version of the plugin. I deleted the caldavzap directory in the Roundcube plugins directory, and added the required info in the composer.json file in Rouncube root.
Here's what I added:

`"offerel/caldavzap": "~1.0.3"`
and ran composer update.
To my surprise it installed version 1.0.4, which is somewhat different from 1.0.3, because now I need to enter configuration in config.js in the caldavzap_0.13.2 directory.
However, in this file I cannot find any references to the server (localhost) or the caldav_path......

Looking at the css console I can see that something is disabling the button : 
`<button class="button-calendar" type="button" id="rcmbtn100" onclick="return rcmail.command('switch-task','caldavzap',this,event)" value="Calendar" disabled="disabled">Calendar</button>`

If this is not in your code, then it must come from Roundcube, or indeed another plugin.

Can you confirm please : 
1) That the disabled part is not in your code ?
2) That 1.0.4 is already available, and where to configure caldav_server and caldav_path ?

Thanks again,
Hans

**Comment by hanslammerts** [comment]
Oh, and I do user larry skin......

**Comment by Offerel** [comment]
The latest version of the plug-in is 1.1.2.

 I don't know why composer try to install the older version. Personally I don't use composer, i install all plugin by directly checking out the version from GitHub. I do this also for other plugins, not only mine.

In the current plugin definition, there is no class "disabled" defined. But maybe some other plugin do add css classes to buttons. That's why its worth to disable for testing purposes all other plugins and re-enable them plug-in for plug-in afterwards.

**Comment by Offerel** [comment]
Ahh and you shouldn't configure anything in config.js. this file isn't used by the plug-in since 1.1.x all configuration is done in config.inc.php

**Comment by hanslammerts** [comment]
Ok, thanks.

I'll delete the caldavzap installation directory again, and retry as I did before using Github.
Also I'll try to disable any other plugins to see if this helps.

Thanks for now. I'll be back !!

**Comment by Offerel** [comment]
I have made now a new release package here at GitHub with all the latest minor fixes. There version number in the Roundcube plugin repository was updated automatically afterwards, but there a little bit strange numbering appears.

A old version version is listed as v1.0.3 and another one as 1.0.3 without the v in front. I don't know why this happens. But also when i delete this v-version, it re-appears again.

I hope the new release will work for you.

**Comment by hanslammerts** [comment]
Ok, I'm back....

Sorry to say that it still does not work.
I completely reinstalled Roundcube, and the underlying database.
Then, I downloaded the latest version you created today, uncompressed it in the plugins folder, copied and edited the conf.inc.php file to reflect the caldav_host and caldav_path, and enabled the plugin in the global conf.inc.php file.
However, the button in the top marked "Calendar" is still disabled. Even when other plugins have been disabled.
When I look in Roundcube under "About" in Roundcube, I can see that only plugins caldavzap (1.1.2), filesystem_attachments (1.0) and jqueryui (1.12.0) have been enabled. I don't see a way to disable the second and third, but I believe I shouldn't.

There is absolutely nothing in the Apache errorlog, and when loading the Roundcube page, all requests are OK (200). There's also nothing in the roundcube/logs/error.log.......

In other words: I'm still stuck and have no clue where this goes wrong.
Do you have any suggestions, please ?

**Comment by Offerel** [comment]
I will look into this today. currently i have no idea why the button should be disabled. In the meantime you can call the plug-in manually, if you add tou your

**Comment by hanslammerts** [comment]
Ok, thanks.

For your information:
Calling the page with ?_task=caldavzap produces a couple of errors:

`ReferenceError: onLoadHandler is not defined
[Learn More]
roundcubemail:1:1`
`Password fields present on an insecure (http://) iframe. This is a security risk that allows user login credentials to be stolen.
[Learn More]
roundcube.html`
`Info: loadConfig
main_rc.js:858:2`
`ReferenceError: onLoadHandler is not defined
[Learn More]
roundcubemail:1:1`
`XML Parsing Error: not well-formed`
`Location: 
Line Number 1, Column 118:`
`Error: [DAVresourceDelegation: 'REPORT http://hans@localhost/davical/caldav.php/hans/'] code: '0' status: 'error' (this error code usually means network connection error, or your browser is trying to make a cross domain query, but it is not allowed by the destination server or the browser itself)
webdav_protocol.js:675:5`

The URL that connects to Davical seems to be wrong.
1) I would expect hans:password@localhost
2) The correct URI should be ../davical/caldav.php/hans/calendar

Hope this info is of any use.

Thanks,
Hans

**Comment by Offerel** [comment]
In the meantime, I have some new insights. The missing onLoadHandler is my fault. It's a remnant of an old version, I just removed it. Error messages related to this should no longer appear.

The thing with the `insecure (http://) iframe`, in your case, is a hint that you only access the URL via http and not https. I assume that your calendar will not be accessible from outside localhost without SSL. In this case, you can ignore it. Otherwise, I would strongly advise you to use an SSL certificate.

`Info: loadConfig main_rc. js: 858:2` is only the info that the primary config file for CalDAVZap was initialized clean.

`XML Parsing Error: not well-formed` you can ignore. This comes from CalDAVZap and would have to be fixed upstream. That's not how deep I am in CalDAVZap. I have these mistakes on me too, but it still works.

It becomes interesting again from `Error:[DAVresourceDelegation:`. CalDAVZap is quite sensitive regarding cross domain. In theory, that should work. However, it seems that you recognize it as a crossdomain. I would try to use the same domain name under which Roundcube is available. At least here it works out very well. With Crossdomain I've had it up and running with Crossdomain before, but that's so much freak with some confusing restrictions that I can't really recommend it.

If you say...` /davical/caldav. php/hans/calendar` would be the correct calendar path, I assume that' calendar' in this case is the name of the main calendar? If so, this specification would be correct. In this case, the complaint comes from the REPORT of Calendar Server. Somehow Davical seems to be expecting a different path. So far I can only speak for SabreDAV and NextCloud. Here, the paths are usually `https://domain.server.de/sabredav/server.php/calendars/<username>/<calendarname>`. The matching entries in the config. php then:

```
$config[' caldav_server'] =' domain. server. de';
$config[' caldav_path'] =' /sabredav/server. php/calendars/';
```

CalDAVZap then extracts the user name and calendar names from the http request's REPORT and merges the path.

Afterwards I commit the remote reference to the onLoadHandler, then I see why maybe the button disabled is. So far I couldn't find anything in the plugin.

**Comment by Offerel** [comment]
Regarding the disabled button, i found the following comment about how to buttons are defined in the Roundcube's skin markup language

> A button is also registered in the UI manager and is automatically enabled and disabled according to the current state of the application.

Since there is no disable function in this plugin (i have searched again), i think something in Roundcube has disabled this button. Since i don't have assigned a special id to this button, the button id is given automatically to this button. In my case this id="rcmbtn100". Maybe this id is used twice somewhere, and this other id must be disabled for other reasons, so it could be the case, that this button is deactivated also by mistake. I can try to give my button a unique id, so that such events do not occur. Since i don't have the disabled button, i can test this. But you can try this approach with the following. There is a file caldavzap.php in the root of the plugin folder. On line 23 begins the button defintion. You can replace this defintion from:

```
$this->add_button(array(
			'label'      => 'caldavzap.caldavzap',
			'command'    => 'caldavzap',
			'class'      => 'button-calendar',
			'classsel'   => 'button-calendar button-selected',
			'innerclass' => 'button-inner'
		), 'taskbar');
```
with:

```
$this->add_button(array(
			'label'      => 'caldavzap.caldavzap',
			'id'		=> 'calendar-button-unique-id',
			'command'    => 'caldavzap',
			'class'      => 'button-calendar',
			'classsel'   => 'button-calendar button-selected',
			'innerclass' => 'button-inner'
		), 'taskbar');
```

and tell me, if this works. The string for the id isn't really important, but it should be a unique one.

**Comment by hanslammerts** [comment]
OK, we're getting somewhere !

First: the adding of an id does not make any difference. The button still remains disabled.
Second: I changed the caldav_host from _localhost_ to the ip address of the machine.
Now, if I call ?_task=caldavzap, I do see the calendar !

It's in German, and empty for now, but that's ok. I don't think I loaded any events yet in Davical...
Since there seems to be functionality in Roundcube to automatically enable and disable applications, I'll also post a question in the Roundcube forum to see if anyone there knows what could be happening here.
And yes, I knew about the http security risk. Don't worry, eventually, when everything works, the entire site will be behind https.

So, for now I think the best thing to do is to wait for an answer to my question over on the Roundcube forum. Do you agree ?

Vielen Dank !
Hans

**Comment by Offerel** [comment]
Immer wieder gern. Schauen wir ob eventuell einer der Devs's im Forum etwas genaueres weiß.

**Comment by Offerel** [comment]
Does it work if you add a type parameter like this:

```
$this->add_button(array(
			'label'		=> 'caldavzap.caldavzap',
			'id'		=> '7ea3a1d1-8270-4985-aaaa-c8227e551e43',
			'command'	=> 'caldavzap',
			'class'		=> 'button-calendar',
			'classsel'	=> 'button-calendar button-selected',
			'innerclass'=> 'button-inner',
			'type'		=> 'link'
		), 'taskbar');
```

**Comment by hanslammerts** [comment]
YES !!!
It does.
The button is not only enabled, but also is shown in the same way like the toolbar options Mail, Contacts and Settings are displayed, that is: with a nice icon displaying it is a calendar. And more importantly, clicking on it really brings me to the calendar iframe....

Great success !

I do know something about PHP programming. Would you care to explain what lead you to this solution, please ?

Danke,
Hans

**Comment by Offerel** [comment]
Ih... Im not really a professional php developer, I have only written this plugin, since no other calendar plugin works as expected. I have only looked in other plugins how the things are done.

Some had the button type defined, some not. since i have seen the type definition in all official plug-ins, it was worth a try. 

Good to know that it works. I will solve this in my other plugins to and release a new version.

**Comment by Offerel** [comment]
Fixed with https://github.com/Offerel/roundcube_caldavzap/commit/39e436ce95266e3ad9664a685c9c85f78ca90be0
